### ISSUE TITLE
Multiple transitions in invoked state machines.

### ISSUE THREAD

**Issue by nathanhammond** [davidkpiano/xstate#1397]
**Description**
Since 460e831, state transitions (when descendant from an `invoke`d `StateNode`) will occur the same number of times as the `StateNode` is nested.

**Expected Result**
The transition occurs exactly once. The `assign` function is called exactly once.

**Actual Result**
The transition occurs twice. The `assign` function is called twice. (The number of invocations is dependent upon its depth in the hierarchy.)

**Reproduction**
These two machines are 100% identical, the only change is the `xstate` dependency version:
- [Working - 4.10](https://codesandbox.io/s/optimistic-nash-tx7vj)
- [Issue - 4.11](https://codesandbox.io/s/sweet-shannon-dvzox)

**Comment by nathanhammond** [comment]
### Bisecting

Bisect steps, in case you wish to reproduce. (Including a diff to make it build on one of the steps.)

```patch
commit 5813a5933a08acd54cf95b2fc1027d7ef4647135
Merge: cc65d1c7 d0aa23f5
Author: Nathan Hammond <github.com@nathanhammond.com>
Date:   Thu Aug 20 21:19:03 2020 +0800

    WIP on (no branch): cc65d1c7 Refactor useConstant calls

diff --cc packages/core/src/actions.ts
index f44e66e2,f44e66e2..12e08599
--- a/packages/core/src/actions.ts
+++ b/packages/core/src/actions.ts
@@@ -106,7 -106,7 +106,6 @@@ export function toActionObject<TContext
        const { type, ...other } = action;
  
        actionObject = {
--        type,
          ...exec,
          ...other
        } as ActionObject<TContext, TEvent>;
diff --cc packages/core/src/interpreter.ts
index ac28865c,ac28865c..39ab9748
--- a/packages/core/src/interpreter.ts
+++ b/packages/core/src/interpreter.ts
@@@ -259,7 -259,7 +259,7 @@@ export class Interpreter
      }
    }
    private update(
--    state: State<TContext, TEvent, TStateSchema, TTypestate>,
++    state: State<TContext, TEvent, TStateSchema, any>,
      _event: SCXML.Event<TEvent>
    ): void {
      // Attach session ID to state
```

```sh
git clone git@github.com:davidkpiano/xstate.git
cd xstate
yarn
cd packages/core && yarn link && cd ../../
git bisect start
git bisect good xstate@4.10.0
git bisect bad xstate@4.11.0
cd packages/core/ && git reset --hard && yarn build && cd ../../ # Apply the patch if necessary
```

**Comment by Andarist** [comment]
Thank you for providing a detailed repro case. I will investigate this soon.

**Comment by Andarist** [comment]
This happens because we read the child's initialState here:
https://github.com/davidkpiano/xstate/blob/ab7329eaf849b613b94cb815a7e1575e9c9f7ea9/packages/core/src/Actor.ts#L79
and here:
https://github.com/davidkpiano/xstate/blob/ab7329eaf849b613b94cb815a7e1575e9c9f7ea9/packages/core/src/interpreter.ts#L477

Maybe we could cache the evaluated initialState so this would happen, but I'm not sure right now if that didn't cause a bug in the past somewhere. To fix your problem you should be able to use this:
```
spawn(() => Promise.resolve())
```

**Comment by nathanhammond** [comment]
Mentioning issue #1237 and pull #1239 which are related to this discussion.

It seems like this might be a bit of a whack-a-mole game.

**Comment by davidkpiano** [comment]
Yeah, the fix will have to be avoiding a duplicate read of `.initialState`.

In version 5, this problem is mitigated systematically, because all actors (spawned and invoked) are initialized in a "not started" state, so you might have something like:

```js
// not allowed: spawn.from(createSomePromise(), 'some-id')

spawn.from(() => createSomePromise(), 'some-id');
```

And that will create an actor that needs `someActor.start()` to be called to actually execute the promise. So it's resilient to being read any number of times because it hasn't "started" yet.

**Comment by nathanhammond** [comment]
> because all actors (spawned and invoked) are initialized in a "not started" state

I find that, as a library pattern, this is likely a tremendous improvement in making xstate easier to integrate into other more-opinionated (say, concurrent) interpreters. This is definitely an improvement that will better enable adoption.

As you implied, resilience and mitigation does not necessarily resolve the underlying issue. The approach in v5 will still retain the problem where a lot of people (myself included!) would do something to make the new distinction irrelevant. Whatever the future equivalent of this is:

```js
{
  actions: [
    assign({ promise: spawn(() => new Promise()) }),
    (context) => { context.promise.start(); }
  ]
}
```

Throw that in a nested `initialState` and we're back at square 1.

I'm going to test my additional `if` clause hypothesis today. I'm pretty new to this code base (and very grateful for the pointers, @Andarist) so I'm treating this as a POC. If it doesn't break the existing tests I'll let you know.
